From 874d76f922658c09c2d9c953ecb2dcd3e9999685 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 21 Nov 2023 10:15:15 +0100
Subject: [PATCH 0357/1695] drivers: stm32_tamp: add STM32MP25x platforms
 tampers support

Add the pin map of STM32MP25x platforms tampers. While there, update
the _TAMP_OR bit fields so that it is obvious that the bit field is
platform dependent.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 5177b44777c28b25fcf22cb3b5896c9646af6164 drivers: stm32_tamp: add STM32MP25x platforms tampers support]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I42c29fe1b8c41aeb5df2185efc89f9dce9164c70
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/343123
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_tamp.c | 80 +++++++++++++++++++++++++++++++++++++--
 1 file changed, 76 insertions(+), 4 deletions(-)

diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index cb041a03c..7d66a8320 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -177,6 +177,13 @@
 #define _TAMP_STM32MP15_OR_OUT3RMP_PI8		U(0)
 #define _TAMP_STM32MP15_OR_OUT3RMP_PC13		BIT(0)
 
+#define _TAMP_STM32MP25_OR_IN1RMP_PC4		U(0)
+#define _TAMP_STM32MP25_OR_IN1RMP_PI8		BIT(0)
+#define _TAMP_STM32MP25_OR_IN3RMP_PC3		U(0)
+#define _TAMP_STM32MP25_OR_IN3RMP_PZ2		BIT(1)
+#define _TAMP_STM32MP25_OR_IN5RMP_PF6		U(0)
+#define _TAMP_STM32MP25_OR_IN5RMP_PZ4		BIT(2)
+
 /* _TAMP_ERCFGR bit fields */
 #define _TAMP_ERCFGR_ERCFG0		BIT(0)
 #define _TAMP_ERCFGR_ERCFG_MASK		BIT(0)
@@ -293,6 +300,25 @@ static const char * const itamper_name[] = {
 DECLARE_KEEP_PAGER(itamper_name);
 #endif
 
+#ifdef CFG_STM32MP25
+static const char * const itamper_name[] = {
+	[INT_TAMP1] = "Backup domain voltage threshold monitoring",
+	[INT_TAMP2] = "Temperature monitoring",
+	[INT_TAMP3] = "LSE monitoring",
+	[INT_TAMP4] = "HSE monitoring",
+	[INT_TAMP5] = "RTC Calendar overflow",
+	[INT_TAMP6] = "JTAG/SWD access",
+	[INT_TAMP7] = "VDDCORE monitoring under/over voltage",
+	[INT_TAMP8] = "Monotonic counter 1 overflow",
+	[INT_TAMP9] = "Cryptographic peripherals fault",
+	[INT_TAMP10] = "Monotonic counter 2 overflow",
+	[INT_TAMP11] = "IWDG3 reset",
+	[INT_TAMP12] = "VDDCPU monitoring under/over voltage",
+	[INT_TAMP14] = "IWDG5_reset",
+	[INT_TAMP15] = "IWDG1_reset",
+};
+#endif
+
 static struct stm32_tamp_conf int_tamp_mp13[] = {
 	{ .id = INT_TAMP1 }, { .id = INT_TAMP2 }, { .id = INT_TAMP3 },
 	{ .id = INT_TAMP4 }, { .id = INT_TAMP5 }, { .id = INT_TAMP6 },
@@ -306,6 +332,15 @@ static struct stm32_tamp_conf int_tamp_mp15[] = {
 	{ .id = INT_TAMP4 }, { .id = INT_TAMP5 }, { .id = INT_TAMP8 },
 };
 
+static struct stm32_tamp_conf int_tamp_mp25[] = {
+	{ .id = INT_TAMP1 }, { .id = INT_TAMP2 }, { .id = INT_TAMP3 },
+	{ .id = INT_TAMP4 }, { .id = INT_TAMP5 }, { .id = INT_TAMP6 },
+	{ .id = INT_TAMP7 }, { .id = INT_TAMP8 }, { .id = INT_TAMP9 },
+	{ .id = INT_TAMP10 }, { .id = INT_TAMP11 },
+	{ .id = INT_TAMP12 }, { .id = INT_TAMP14 },
+	{ .id = INT_TAMP15 },
+};
+
 static struct stm32_tamp_conf ext_tamp_mp13[] = {
 	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
 	{ .id = EXT_TAMP4 }, { .id = EXT_TAMP5 }, { .id = EXT_TAMP6 },
@@ -316,6 +351,12 @@ static struct stm32_tamp_conf ext_tamp_mp15[] = {
 	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
 };
 
+static struct stm32_tamp_conf ext_tamp_mp25[] = {
+	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
+	{ .id = EXT_TAMP4 }, { .id = EXT_TAMP5 }, { .id = EXT_TAMP6 },
+	{ .id = EXT_TAMP7 }, { .id = EXT_TAMP8 },
+};
+
 #define GPIO_BANK(port)	 ((port) - 'A')
 
 static const struct stm32_tamp_pin_map pin_map_mp13[] = {
@@ -380,6 +421,33 @@ static const struct stm32_tamp_pin_map pin_map_mp15[] = {
 	},
 };
 
+static const struct stm32_tamp_pin_map pin_map_mp25[] = {
+	{
+		.id = EXT_TAMP1, .bank = GPIO_BANK('I'), .pin = 8,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN1RMP_PI8,
+	},
+	{
+		.id = EXT_TAMP1, .bank = GPIO_BANK('C'), .pin = 4,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN1RMP_PC4,
+	},
+	{
+		.id = EXT_TAMP3, .bank = GPIO_BANK('C'), .pin = 3,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN3RMP_PC3,
+	},
+	{
+		.id = EXT_TAMP3, .bank = GPIO_BANK('Z'), .pin = 2,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN3RMP_PZ2,
+	},
+	{
+		.id = EXT_TAMP5, .bank = GPIO_BANK('F'), .pin = 6,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN5RMP_PF6,
+	},
+	{
+		.id = EXT_TAMP5, .bank = GPIO_BANK('Z'), .pin = 4,
+		.out = false, .conf = _TAMP_STM32MP25_OR_IN5RMP_PZ4,
+	},
+};
+
 /* Expects at most a single instance */
 static struct stm32_tamp_instance stm32_tamp;
 
@@ -663,11 +731,9 @@ static TEE_Result stm32_tamp_set_int_config(struct stm32_tamp_compat *tcompat,
 			*cr3 &= ~_TAMP_CR3_ITAMPNOER(id);
 	}
 
-#ifndef CFG_STM32MP25
 	DMSG("'%s' internal tamper enabled in %s mode",
 	     itamper_name[id - INT_TAMP1],
 	     (tamp_int->mode & TAMP_NOERASE) ? "potential" : "confirmed");
-#endif
 
 	return TEE_SUCCESS;
 }
@@ -1612,12 +1678,11 @@ stm32_tamp_parse_passive_conf(const void *fdt, int node,
 static uint32_t stm32_tamp_itamper_action(int __maybe_unused id)
 {
 	const char __maybe_unused *tamp_name = NULL;
-#ifndef CFG_STM32MP25
+
 	if (id >= 0 && ((size_t)id < ARRAY_SIZE(itamper_name)))
 		tamp_name = itamper_name[id];
 
 	MSG("Internal tamper %u (%s) occurs", id - INT_TAMP1 + 1, tamp_name);
-#endif
 
 	return TAMP_CB_ACK_AND_RESET;
 }
@@ -2044,6 +2109,7 @@ static const struct stm32_tamp_compat mp15_compat = {
 };
 
 static const struct stm32_tamp_compat mp25_compat = {
+		.nb_monotonic_counter = 2,
 		.tags = TAMP_HAS_REGISTER_SECCFGR |
 			TAMP_HAS_REGISTER_PRIVCFGR |
 			TAMP_HAS_RIF_SUPPORT |
@@ -2051,6 +2117,12 @@ static const struct stm32_tamp_compat mp25_compat = {
 			TAMP_HAS_REGISTER_CR3 |
 			TAMP_HAS_REGISTER_ATCR2 |
 			TAMP_HAS_CR2_SECRET_STATUS,
+		.int_tamp = int_tamp_mp25,
+		.int_tamp_size = ARRAY_SIZE(int_tamp_mp25),
+		.ext_tamp = ext_tamp_mp25,
+		.ext_tamp_size = ARRAY_SIZE(ext_tamp_mp25),
+		.pin_map = pin_map_mp25,
+		.pin_map_size = ARRAY_SIZE(pin_map_mp25),
 };
 
 static const struct dt_device_match stm32_tamp_match_table[] = {
-- 
2.47.2

